src/cluster/websocket/WebSocket.cc

```diff
@@ -0,0 +1,487 @@
+        return;
+    }
+
+    subscriptions_state[topic_prefix] = true;
+}
+
+bool WebSocketClient::AllSubscriptionsActive() const {
+    for ( const auto& [_, status] : subscriptions_state ) {
+        if ( ! status )
+            return false;
+    }
+
+    return true;
+}
+
+const std::vector<std::string> WebSocketClient::GetSubscriptions() const {
+    std::vector<std::string> subs;
+    subs.reserve(subscriptions_state.size());
+
+    for ( const auto& [topic, _] : subscriptions_state )
+        subs.emplace_back(topic);
+
+    return subs;
+}
+
+class zeek::cluster::websocket::detail::ReplyMsgThread : public zeek::threading::MsgThread {
+public:
+    ReplyMsgThread() : zeek::threading::MsgThread() { SetName("ws-reply-thread"); }
+
+    void Run() override {
+        zeek::util::detail::set_thread_name("zk/ws-reply-thread");
+        MsgThread::Run();
+    }
+
+    bool OnHeartbeat(double network_time, double current_time) override { return true; }
+
+    bool OnFinish(double network_time) override { return true; }
+};
+
+WebSocketEventDispatcher::WebSocketEventDispatcher() {
+    onloop =
+        new zeek::detail::OnLoopProcess<WebSocketEventDispatcher, WebSocketEvent>(this, "WebSocketEventDispatcher");
+    // Register the onloop instance the IO loop. Lifetime will be managed by the loop.
+    onloop->Register(false);
+
+    reply_msg_thread = new ReplyMsgThread();
+    reply_msg_thread->Start();
+}
+
+WebSocketEventDispatcher::~WebSocketEventDispatcher() {
+    // Freed by threading manager.
+    reply_msg_thread = nullptr;
+}
+
+void WebSocketEventDispatcher::Terminate() {
+    WS_DEBUG("Terminating WebSocketEventDispatcher");
+
+    for ( auto& [_, client] : clients ) {
+        const auto& wsc = client.wsc;
+        const auto& backend = client.backend;
+        WS_DEBUG("Sending close to WebSocket client %s (%s:%d)", wsc->getId().c_str(), wsc->getRemoteIp().c_str(),
+                 wsc->getRemotePort());
+
+        QueueReply(WebSocketCloseReply{wsc, 1001, "Terminating"});
+        backend->Terminate();
+    }
+
+    clients.clear();
+}
+
+void WebSocketEventDispatcher::QueueForProcessing(WebSocketEvent&& event) {
+    // Just delegate to onloop. The work will be done in Process()
+    onloop->QueueForProcessing(std::move(event));
+}
+
+void WebSocketEventDispatcher::QueueReply(WebSocketReply&& reply) {
+    // Delegate to the reply thread.
+    reply_msg_thread->SendIn(new ReplyInputMessage(std::move(reply)));
+}
+
+// WebSocketDemux::Process() runs in the main thread.
+//
+// XXX: How is this going to work with class broker? With
+// ZeroMQ, each WebSocket client has its own XPUB/XSUB
+// connectivity to a central broker and similarly with NATS.
+// But with broker we need to do something different.
+// Maybe connect to the local endpoint.
+//
+// We cannot actually instantiate a Broker backend :-(
+//
+// We could also have InitPostScript() recognize Broker
+// and start its internal server instead.
```

Comment:
I'm not following this, let's talk this over in person. I thought that in the next iteration of the refactoring Broker becomes a backend like the others, but I'm probably just missing the point.

Reply:
The `InitPostScript()` part was about switching `Cluster::listen_websocket()` to be `Broker::listen_websocket()` instead, but I should just be bold enough to make the native Zeek level WebSocket work properly.

This would then also allows to change the JSON format for WebSocket clients, even if Broker is used as a Backend.

In summary, after this PR, `Cluster::listen_websocket()` with broker as cluster backend will not be functional, but there's a sketch of a way forward.

Reply:
Thanks for the explanation and walking me through it in person today, I get it now.